package/src/hydra-ai/ai-service.ts

```diff
@@ -119,51 +125,103 @@ This response should be short and concise.`;
         throw new Error(`Component ${componentName} not found`);
       }
 
-      const generateComponentPrompt = `You are an AI assistant that interacts with users and helps them perform tasks.
+      return this.hydrateComponent(context.prompt, component);
+    } else {
+      // TODO: Handle this case. Maybe repeat the decision prompt.
+      throw new Error("Invalid decision");
+    }
+  };
+
+  async hydrateComponent(
+    message: string,
+    component: AvailableComponent,
+    toolResponse?: any
+  ): Promise<ComponentDecision> {
+    const generateComponentPrompt = `You are an AI assistant that interacts with users and helps them perform tasks.
 To help the user perform these tasks, you are able to generate UI components. You are able to display components and decide what props to pass in. However, you can not interact with, or control 'state' data.
 When prompted, you will be told the component to display, a description of any props to pass in, and any other related context.
 You will also be given a user message. Use the user message and the provided context to determine what props to pass in.
+${
+  toolResponse
+    ? `You have received a response from a tool. Use this data to help determine what props to pass in: ${JSON.stringify(
+        toolResponse
+      )}`
+    : `You can also use any of the provided tools to fetch data needed to pass into the component.`
+}
 
 ${this.generateZodTypePrompt(schema)}`;
 
-      const generateComponentResponse = await this.callOpenAI(
-        [
-          {
-            role: "system",
-            content: generateComponentPrompt,
-          },
-          {
-            role: "user",
-            content: `<componentName>${componentName}</componentName>
+    const tools = toolResponse
+      ? undefined
+      : this.openAIToolFromMetadata(component.contextTools);
+
+    const generateComponentResponse = await this.callOpenAI(
+      [
+        {
+          role: "system",
+          content: generateComponentPrompt,
+        },
+        {
+          role: "user",
+          content: `<componentName>${component.name}</componentName>
 <expectedProps>${JSON.stringify(component.props)}</expectedProps>
-<context>${JSON.stringify(component.context)}</context>
-<userMessage>${context.prompt}</userMessage>`,
-          },
-        ], 
-        true
+<userMessage>${message}</userMessage>
+${
+  toolResponse && `<toolResponse>${JSON.stringify(toolResponse)}</toolResponse>`
+}`,
+        },
+      ],
+      tools,
+      true
+    );
+
+    const componentDecision: ComponentDecision = {
+      message: "Fetching additional data",
+      componentName: component.name,
+      props: null,
+      toolCallRequest: generateComponentResponse.toolCallRequest,
+    };
+
+    if (generateComponentResponse.message !== "") {
```

Comment:
is `response.message` guaranteed to be empty if it call a tool?

might be a better way to check/separate the cases.